### PR TITLE
Add promise state inspection methods

### PR DESCRIPTION
**Pull request by jsor**
Extends the `PromiseInterface` with the state inspection methods `isFulfilled()`, `isRejected()`, `isPending()`, `isCancelled()`, `value()` and `reason()`.

This especially allows for optimizations in functions working on promises directly. For example, [`React\Promise\Timer\timeout`](https://github.com/reactphp/promise-timer/blob/master/src/functions.php#L10) can short circuit and avoid registering a timer if the promise is already resolved.

```php
function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
{
    if (!$promise->isPending()) {
        return $promise;
    }

    // ...
}
```

### CODE REVIEW

**Review by WyriHaximus — APPROVED** [review]

File: src/Promise.php [hunk 1]
```diff
@@ -76,6 +83,12 @@ public function always(callable $onFulfilledOrRejected)
 
     public function cancel()
     {
+        if (null !== $this->result) {
```
Comment (kelunik) [inline src/Promise.php]: What if a promise is resolved with `null`?

Comment (jsor) [inline src/Promise.php]: `$this->result` is always a `PromiseInterface` instance after resolution.

**Comment by kelunik** [comment]
Something like this might result in problems with `LazyPromise` in case async resolution gets implemented.

**Comment by jsor** [comment]
@kelunik Could you elaborate?

**Comment by kelunik** [comment]
A promise that's already resolved will return `true`, while a `LazyPromise` will return `false` and only return `true` in the next tick in case of async resolution and the `LazyPromise` immediately resolving. On the other hand, constant / immediately available values are probably not wrapped in a `LazyPromise`.

**Comment by clue** [comment]
I'm not entirely sold on this idea yet, can you elaborate on what's the use-case for these methods and why I wouldn't use the existing methods? Thanks!

**Comment by jsor** [comment]
It mostly allows for optimizations. One use case is mentioned in the PR description. [Here's](http://bluebirdjs.com/docs/api/synchronous-inspection.html) another from bluebird docs.

**Comment by WyriHaximus** [comment]
@clue here is a usecase: https://github.com/reactphp/promise-timer/pull/25 :stuck_out_tongue:

**Comment by clue** [comment]
@WyriHaximus here is why I'm not sold on the idea that these methods are needed: reactphp/promise-timer#25 :stuck_out_tongue:

---

Now seriously, I'm not saying they *aren't* useful, but I'm not seeing *why* they are useful either. Perhaps we *may* want to add some documentation that highlights *why* they are useful in the first place? In particular from a consumer perspective, when should I use which methods?

> *Premature optimization is the root of all evil – Donald Knuth*

IMHO we should carefully consider if increasing our API surface by 100% is really worth it :+1:

**Comment by jsor** [comment]
I can't see premature optimization here. We're not optimizing, we're providing tools for optimizations. The promise-timer issue and the multiline workaround you implemented (introducing side issues like swallowing exceptions) clearly shows the need for it.

**Comment by WyriHaximus** [comment]
@clue touché

To offer an alternative: what about a method state that returns one of the four states rather then add 6 methods?

```php
/**
 * Returns either:
 * 'pending'
 * 'resolved'
 * 'rejected'
 * 'canceled'
 */
public function state(): string;
```

P.S. I'm always in favour for more/better documentation, even though it isn't my strongest skill.

**Comment by jsor** [comment]
@WyriHaximus I don't see a benefit because that only moves explicit code to documentation and makes no difference from an implemention point (implement 4 states in `state()` vs. implement 4 explicit methods). Also, the `value()` and `reason()` getters are now missing.

It also makes it more error-prone: you already misspelled _cancelled_ in your example ;)

**Comment by WyriHaximus** [comment]
@jsor I don't have a strong opinion on this matter tbh, I do like the feature :+1: . (Left `value()` and `reason()` out because I didn't thought up an alternative for them.)

> It also makes it more error-prone: you already misspelled cancelled in your example ;)

I'm a TaaS, Typos as a Service :stuck_out_tongue_closed_eyes:

**Pull request closed without merging**